core/sys/orca/orca.odin

```diff
@@ -0,0 +1,2183 @@
+	ellipse_fill :: proc(x: f32, y: f32, rx: f32, ry: f32) ---
+	ellipse_stroke :: proc(x: f32, y: f32, rx: f32, ry: f32) ---
+	circle_fill :: proc(x: f32, y: f32, r: f32) ---
+	circle_stroke :: proc(x: f32, y: f32, r: f32) ---
+	arc :: proc(x: f32, y: f32, r: f32, arcAngle: f32, startAngle: f32) ---
+	text_fill :: proc(x: f32, y: f32, text: str8) ---
+	image_draw :: proc(image: image, rect: rect) ---
+	image_draw_region :: proc(image: image, srcRegion: rect, dstRegion: rect) ---
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// A surface for rendering using the GLES API.
+////////////////////////////////////////////////////////////////////////////////
+
+@(default_calling_convention="c", link_prefix="oc_")
+foreign {
+	gles_surface_create :: proc() -> surface ---
+	gles_surface_make_current :: proc(surface: surface) ---
+	gles_surface_swap_interval :: proc(surface: surface, interval: i32) ---
+	gles_surface_swap_buffers :: proc(surface: surface) ---
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Graphical User Interface API.
+////////////////////////////////////////////////////////////////////////////////
+
+key_state :: struct {
+	lastUpdate: u64,
+	transitionCount: u32,
+	repeatCount: u32,
+	down: c.bool,
+	sysClicked: c.bool,
+	sysDoubleClicked: c.bool,
+	sysTripleClicked: c.bool,
+}
+
+keyboard_state :: struct {
+	keys: [349]key_state,
+	mods: keymod_flags,
+}
+
+mouse_state :: struct {
+	lastUpdate: u64,
+	posValid: c.bool,
+	pos: vec2,
+	delta: vec2,
+	wheel: vec2,
+	_: struct #raw_union {
+		buttons: [5]key_state,
+		_: struct {
+			left: key_state,
+			right: key_state,
+			middle: key_state,
+			ext1: key_state,
+			ext2: key_state,
+		},
+	},
+}
+
+BACKING_SIZE :: 64
+
+text_state :: struct {
+	lastUpdate: u64,
+	backing: [64]utf32,
+	codePoints: str32,
+}
+
+clipboard_state :: struct {
+	lastUpdate: u64,
+	pastedText: str8,
+}
+
+input_state :: struct {
+	frameCounter: u64,
+	keyboard: keyboard_state,
+	mouse: mouse_state,
+	text: text_state,
+	clipboard: clipboard_state,
+}
+
+ui_key :: struct {
+	hash: u64,
+}
+
+ui_axis :: enum u32 {
+	X = 0,
+	Y = 1,
+	COUNT = 2,
+}
+
+ui_align :: enum u32 {
+	START = 0,
+	END = 1,
+	CENTER = 2,
+}
+
+ui_layout_align :: [2]ui_align
+
+ui_layout :: struct {
+	axis: ui_axis,
+	spacing: f32,
+	margin: [2]f32,
+	align: ui_layout_align,
+}
+
+ui_size_kind :: enum u32 {
+	TEXT = 0,
+	PIXELS = 1,
+	CHILDREN = 2,
+	PARENT = 3,
+	PARENT_MINUS_PIXELS = 4,
+}
+
+ui_size :: struct {
+	kind: ui_size_kind,
+	value: f32,
+	relax: f32,
+	minSize: f32,
+}
+
+ui_box_size :: [2]ui_size
+
+ui_box_floating :: [2]bool
+
+ui_style :: struct {
+	size: ui_box_size,
+	layout: ui_layout,
+	floating: ui_box_floating,
+	floatTarget: vec2,
+	_color: color,
+	bgColor: color,
+	borderColor: color,
+	font: font,
+	fontSize: f32,
+	borderSize: f32,
+	roundness: f32,
+	animationTime: f32,
+	animationMask: ui_style_mask,
+}
+
+ui_palette :: struct {
```

Comment:
Random comment regarding Orca itself, but this struct is kind of bizarre to me why it's not just a struct containing an array instead.

Reply:
ui library feedback is welcome, same with their current styling api